Add support for EFM32GG11

stevew817 opened the pull request:
### Description

Adding in support for EFM32GG11 - STK3701 (https://www.silabs.com/products/development-tools/mcu/32-bit/efm32-giant-gecko-gg11-starter-kit) as a target, now that the EMAC changes have made their way to master.

#### What's supported with this PR
* Core mbed-os support (ADC/serial/SPI/I2C/tickless/PWM/GPIO/RTC/lowpower timer)
* mbedTLS hardware acceleration support (AES-128/256, SHA-1/2, ECC secp192/224/256, TRNG)
* Ethernet (EMAC driver)

#### Missing board features
* SD slot is not supported, since it is hardwired to the chip's SDIO peripheral. mbed-os does not support SDIO AFAIK.
* On-board SPI flash is hardwired to the chip's QSPI peripheral. This functionality needs the QSPI HAL in order to be ported. @0xc0170 do you have a timeline for that addition?

#### Test results
[Test results](https://www.dropbox.com/s/6lj0zjnqtw40cga/EFM32GG11-mbed-testresults.txt?dl=0)

### Pull request type

    [ ] Fix
    [ ] Refactor
    [X] New target
    [ ] Feature
    [ ] Breaking change

stevew817 committed:
Add EFM32GG11_STK3701 support

stevew817 committed:
Add EFM32GG11 Ethernet driver

0xc0170 commented:
Travis doxygen failure is related to the docs , please review.

stevew817 commented:
@0xc0170 fixed :)

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Silicon_Labs/sl_emac_config.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef SL_EMAC_CONFIG_H
```
license missing here

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Silicon_Labs/sl_eth_hw.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef SL_ETH_HW_H
```
missing license here

0xc0170 commented:
> On-board SPI flash is hardwired to the chip's QSPI peripheral. This functionality needs the QSPI HAL in order to be ported. @0xc0170 do you have a timeline for that addition?

cc @bulislaw

bulislaw commented:
Hope we can get it in for 5.10. Let me try again to get the things moving.

stevew817 commented:
@0xc0170 Licenses are now correctly applied

cmonr commented:
Man is this a feature-filled target.

@ARMmbed/mbed-os-hal @ARMmbed/mbed-os-ipcore @ARMmbed/mbed-os-tls 
Thoughts on at least sections of this PR?

0xc0170 commented:
/morph build

0xc0170 commented:
Please review the latest test results - fails to  find `malloc.h` file?

KariHaapalehto commented:
Failing seems to happen with USI_WM_BN_BM_22-target, which is meant to be base target for MTB_USI_WM_BN_BM_22 and MTB_ADV_WISE_1530. 
MTB_USI_WM_BN_BM_22 and MTB_ADV_WISE_1530 are inheriting USI_WM_BN_BM_22,
so USI_WM_BN_BM_22 is not meant to be real target. I'll add the:
"public": false
to target.json

stevew817 commented:
@0xc0170 fixed, malloc.h works on GCC but it's really the stdlib include that we should've used.

0xc0170 commented:
/morph build

cmonr commented:
@stevew817 Much closer, but there's still one build issue.

bulislaw commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG11/device/TOOLCHAIN_ARM_STD/startup_efm32gg11.S:
```diff
@@ -0,0 +1,373 @@
+;   <o> Stack Size (in Bytes) <0x0-0xFFFFFFFF:8>
+; </h>
+
+Stack_Size      EQU     0x00004000
```
That should be 1K, we use it only as ISR/Scheduler stack, I know the logic is convoluted in mbed_boot.c. Refactor and tests are on their way to.

bulislaw commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG11/device/TOOLCHAIN_GCC_ARM/startup_efm32gg11.S:
```diff
@@ -0,0 +1,395 @@
+#ifdef __STACK_SIZE
+    .equ        Stack_Size, __STACK_SIZE
+#else
+    .equ        Stack_Size, 0x00004000
```
That should be 1K, we use it only as ISR/Scheduler stack, I know the logic is convoluted in mbed_boot.c. Refactor and tests are on their way to.

bulislaw commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG11/device/TOOLCHAIN_IAR/efm32gg11b820f2048gl192.icf:
```diff
@@ -0,0 +1,36 @@
+define symbol __ICFEDIT_region_RAM_end__     = 0x2007FFFF;
+/*-Sizes-*/
+/*Heap 1/4 of ram and stack 1/8*/
+define symbol __ICFEDIT_size_cstack__   = 0x4000;
```
That should be 1K, we use it only as ISR/Scheduler stack, I know the logic is convoluted in mbed_boot.c. Refactor and tests are on their way to.

stevew817 commented:
@cmonr I'm not quite sure where the ARM build failure is coming from. The get_default_instance function is defined in sl_emac.cpp, no?

@bulislaw What about mbed classic?

cmonr commented:
@ARMmbed/mbed-os-ipcore Thoughts on where the error might be coming from?

mikaleppanen commented:
Build error seems to relate to lack of feature LWIP in targets.json configuration file for the board. Seppo has removed the LWIP feature from master branch. So the build is likely to work now. @SeppoTakalo could you confirm?

SeppoTakalo commented:
Please rebase on top of master

`FEATURE_LWIP/lwip-interface` is now renamed to `lwip-stack`

Also targets do not need to specify `features_add: ["LWIP"]` anymore.

stevew817 commented:
@cmonr @SeppoTakalo Now rebased on master.

@bulislaw - Can you assure that your proposed changes don't impact mbed classic (w/o RTOS) builds? Those apps usually have a bunch more stuff on their stack than 1K...

bulislaw commented:
Good point, @mprse can you make sure we are not limiting stack to 1K on RTOS-less Mbed.

stevew817 commented:
Is this blocked by #7238? If not, can someone re-kick CI?

0xc0170 commented:
> Is this blocked by #7238? If not, can someone re-kick CI?

Needs bit more work and testing (can be then rebased and this included there I believe). For now, starting CI here

0xc0170 commented:
/morph build

simonbutcher commented:
@0xc0170 - Can you please fix the reviewers? This needs to be reviewed by @ARMmbed/mbed-os-crypto, not @mbed-os-tls.

cc: @Patater

0xc0170 commented:
> @0xc0170 - Can you please fix the reviewers? This needs to be reviewed by @ARMmbed/mbed-os-crypto, not @mbed-os-tls.

Reviewers fixed.

> mbedTLS hardware acceleration support (AES-128/256, SHA-1/2, ECC secp192/224/256, TRNG)

 I checked and there is not much review there (as TRNG is already supported for other targets that share the same implementation). I believe same is true for crypto (could not locate the new implementation). @stevew817 is that correct?

stevew817 commented:
Yes, that's correct. These are the same crypto peripherals that were already ported for earlier EFR targets, so no changes to the crypto code.

yanesca commented:
If there are no changes in cryptographic code, then no review from us is necessary.

stevew817 commented:
@ARMmbed/mbed-os-core Can someone take a look at the exporter failure? I can't figure out where it is coming from, and given that it fails on assembly files (*.S) from both our CMSIS startup and your RTOS core, I'm thinking this is an infrastructure thing...

0xc0170 commented:
/morph export-build

theotherjimmy commented:
@stevew817 @studavekar I think uvision is missing the device support CMSIS Pack.

stevew817 commented:
@theotherjimmy We have released the CMSIS pack for EFM32GG11 as far as I can see. It is both on the uVision/MDK5 pack index (https://www.keil.com/dd2/pack/) and in the mbed-OS JSON (https://raw.githubusercontent.com/ARMmbed/mbed-os/master/tools/arm_pack_manager/index.json)

Am I missing something that I should have updated?

theotherjimmy commented:
@stevew817 The uvision install in our CI. :D Don't worry, @studavekar will look into that for you.

stevew817 commented:
@studavekar How is this coming along?

0xc0170 commented:
I'll check the progress for this

studavekar commented:
Internal tracker : IOTTESTINF-2980

cmonr commented:
Could not check out ARM license.
/morph build

cmonr commented:
Nvm, was not suppose to be started.

stevew817 commented:
@cmonr CI status is really messed up right now. Any chances of getting it fixed and getting the green light on this PR?

cmonr commented:
@stevew817 We're still waiting on the pack to be added into CI.
Meanwhile, I'll chase @bulislaw, since I think his question/response was fogotten about.

stevew817 commented:
@cmonr How's the pack update coming along?

0xc0170 commented:
> @cmonr How's the pack update coming along?

Our CI infra team is handling the request. They should provide info soon about the progress

RomanSaveljev commented:
New agent image has been prepared and deployed. Please test.

0xc0170 commented:
/morph build

0xc0170 commented:
The build failures are related to rtx internal types used (please update to mbed storage types).

stevew817 commented:
Well, that was an unfortunate surprise. I guess having the PR open for this long does make it prone to master being a very moving target.

Rebased to master and fixed. Should be passing now. @cmonr

cmonr commented:
/morph build

0xc0170 commented:
Test errors not related, we will investigate

cmonr commented:
/morph test

0xc0170 commented:
Full rebuild needed

/morph build

cmonr commented:
Damn it. Because this was closed and re-opened, the PR lost the release tag, and we brought it in without one, meaning it missed the 5.9.3 release...

At least now it'll be in 5.9.4.
@ARMmbed/mbed-os-maintainers we missed one :/

amq commented:
@stevew817 do you also plan to add support for CAN?
